src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -11,13 +11,64 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
+    /// <summary>
+    /// A builder for lazy instances of TypeSymbolWithAnnotations.
+    /// </summary>
+    [DebuggerDisplay("{GetDebuggerDisplay(), nq}")]
+    internal struct TypeSymbolWithAnnotationsBuilder
+    {
+        private TypeSymbol _defaultType;
+        private INonNullTypesContext _nonNullTypesContext;
+        private bool _isAnnotated;
+        private TypeSymbolWithAnnotationsImplementation _implementation;
+
+        internal TypeSymbol DefaultType => _defaultType;
+        internal INonNullTypesContext NonNullTypesContext => _nonNullTypesContext;
+        internal bool IsAnnotated => _isAnnotated;
+        internal TypeSymbolWithAnnotationsImplementation Implementation => _implementation;
+
+        internal bool IsNull => _defaultType is null;
+
+        internal bool InterlockedInitialize(TypeSymbolWithAnnotations type)
+        {
+            if ((object)_defaultType != null)
+            {
+                return false;
+            }
+            _isAnnotated = type.IsAnnotated;
+            Interlocked.CompareExchange(ref _nonNullTypesContext, type.NonNullTypesContext, null);
+            Interlocked.CompareExchange(ref _implementation, type.Implementation, null);
```

Comment:
Don't think we can safely use `CompareExchange` here. The fields of a struct can always be assigned directly by way of simple struct assignment. Example

```
TSWABuilder t1 = someOtherTSWABuilder
```

The use of `CompareExchange` here is only safe if we also categorically disable assignment of these builder types.

Reply:
NM: the compare exchange is into the TSWA value. First day back and I haven't finished my coffee yet.

---
In reply to: [206542229](https://github.com/dotnet/roslyn/pull/28943#discussion_r206542229) [](ancestors = 206542229)

Reply:
The builder is intended for use for lazy type calculations in `Symbol` implementations. Those callers are responsible for hiding the underlying field and ensuring access is otherwise thread-safe, as they are today.

---
In reply to: [206542229](https://github.com/dotnet/roslyn/pull/28943#discussion_r206542229) [](ancestors = 206542229)